prime-router/src/main/kotlin/history/azure/ReportFileFunction.kt

```diff
@@ -151,6 +140,61 @@ abstract class ReportFileFunction(
         }
     }
 
+    /**
+     * Check for auth issues when fetching a single result.
+     *
+     * @param request HTML request body.
+     * @param id Either a reportId or actionId to look for matches on.
+     * @return The error response if found, or null if all is well.
+     */
+    internal fun authSingleBlocks(
+        request: HttpRequestMessage<String?>,
+        id: String,
+    ): HttpResponseMessage? {
+        // Do authentication
+        val claims = AuthenticationStrategy.authenticate(request)
+            ?: return HttpUtilities.unauthorizedResponse(request, authenticationFailure)
+
+        logger.info("Authenticated request by ${claims.userName}: ${request.httpMethod}:${request.uri.path}")
+
+        val action = this.actionFromId(id)
+
+        return if (!this.actionIsValid(action))
+            HttpUtilities.notFoundResponse(request, "$id is not a valid report")
+        else if (!reportFileFacade.checkSenderAccessAuthorization(action, claims)) {
+            logger.warn(
+                "Invalid Authorization for user ${claims.userName}:" +
+                    " ${request.httpMethod}:${request.uri.path}"
+            )
+            HttpUtilities.unauthorizedResponse(request, authorizationFailure)
+        } else {
+            currentAction = action
+            logger.info(
+                "Authorized request by ${claims.organizationNameClaim} to read ${action.sendingOrg}/submissions"
+            )
+            null
+        }
+    }
+
+    /**
+     * Look for an action related to the given id.
+     *
+     * @param id Either a reportId or actionId to look for matches on.
+     * @return The action related to the given id.
+     */
+    private fun actionFromId(id: String): Action {
```

Comment:
update kdoc to reduce confusion surrounding 'currentAction' - seomthing like 'if we already have a currentAction and the id is the same as the one we are looking for, use that instead of hitting the database again'.

Reply:
updated